src/local-fs.ts

```diff
@@ -256,6 +260,21 @@ export default class LocalFS implements ILocalFSPackageManager {
     return readTarballStream;
   }
 
+  saveToken(token: Token): Promise<any> {
+    if (this.tokenDb == null) {
+      this.tokenDb = level(this.dbPath);
+    }
+    return this.tokenDb.put(token.key, token);
+  }
+
+  deleteToken(user: string, tokenKey: string): Promise<any> {
```

Comment:
I think I made a mistake here
https://github.com/verdaccio/flow-types/commit/14adfb2d1163a19f4b697aabdba5d1f12870e065#diff-ee19344a45f19c3f7dd6bd12db642287

The `ILocalPackageManager` should handle only there packages (tarball/metadata) in the storage. While Tokens are something completely different. For instance, the `getSecret()` method, might be stored in a different place.

I'm afraid we must move those 3 methods to `ILocalData<T>`.

```
interface ILocalData<T> extends IPlugin<T> {
		logger: Logger;
		config: T & Config;
		add(name: string, callback: Callback): void;
		remove(name: string, callback: Callback): void;
		get(callback: Callback): void;
		getSecret(): Promise<string>;
		setSecret(secret: string): Promise<any>;
		getPackageStorage(packageInfo: string): IPackageStorage;
		search(onPackage: Callback, onEnd: Callback, validateName: Function): void;
		saveToken(token: Token): Promise<any>;
		deleteToken(user: string, tokenKey: string): Promise<any>;
		readTokens(filter: TokenFilter): Promise<Array<Token>>;
	}
```

The next step would be store tokens. Here I have mixed feelings. Not sure whether we use a new file `tokens.json` or something or `verdaccio-db.json`, what do you think?

Reply:
I would like to ask you one question since I didn't dig deeper right now:
the tarballs are associated to a list of groups of users in case of `access` configuration right?
In case, I think tokens are related to this kind of information.
Could we think a way to use tokens as groups? Is it worth?

Reply:
> the tarballs are associated to a list of groups of users in case of access configuration right?

Actually, the permissions are decoupled with the storage. Tarballs and metadata only resources. What is tight with the access permission is the name of the dependency (either if the request is metadata or tarball).

The storage plugin is composed of two entities:

- Private registry database (secret, private packages)
- Storage Manager (tarballs metadata)

From my perspective, I'd save tokens in the private registry area.

Reply:
A bit of context, this architecture was initially embebed within Verdaccio before `v3`. Perhaps in the future, we should take out the `Private registry` part outside of this plugin, for such reason I recommend do not mix the storage with tokens.

Reply:
I have updated the PR.

I should have understand the mission: not couple the `Private registry database (secret, private packages)` with tokens in order to change easily in future how to manage tokens.

If all the pieces are in place I would proceed with the implementation 👍